python/ray/autoscaler/gcp/config.py

```diff
@@ -103,52 +105,83 @@ def generate_rsa_key_pair():
     return public_key, pem
 
 
-def _create_crm(gcp_credentials):
+def _create_crm_service_account(gcp_credentials=None):
     return discovery.build(
         "cloudresourcemanager", "v1", credentials=gcp_credentials)
 
 
-def _create_iam(gcp_credentials):
+def _create_iam_service_account(gcp_credentials=None):
     return discovery.build("iam", "v1", credentials=gcp_credentials)
 
 
-def _create_compute(gcp_credentials):
+def _create_compute_service_account(gcp_credentials=None):
     return discovery.build("compute", "v1", credentials=gcp_credentials)
 
 
-def fetch_gcp_credentials_from_provider_config(provider_config):
+def _create_crm_oauth_token_request(token_request):
+    return discovery.build("cloudresourcemanager", "v1", http=token_request)
+
+
+def _create_iam_oauth_token_request(token_request):
+    return discovery.build("iam", "v1", http=token_request)
+
+
+def _create_compute_oauth_token_request(token_request):
+    return discovery.build("compute", "v1", http=token_request)
+
+
+def construct_clients_from_provider_config(provider_config):
     """
     Attempt to fetch and parse the JSON GCP credentials from the provider
     config yaml file.
     """
-    service_account_info_string = provider_config.get("gcp_credentials")
-    if service_account_info_string is None:
+    gcp_credentials = provider_config.get("gcp_credentials")
+    if gcp_credentials is None:
         logger.info("gcp_credentials not found in cluster yaml file. "
                     "Falling back to GOOGLE_APPLICATION_CREDENTIALS "
                     "environment variable.")
         # If gcp_credentials is None, then discovery.build will search for
         # credentials in the local environment.
-        return None
-
-    # If parsing the gcp_credentials failed, then the user likely made a
-    # mistake in copying the credentials into the config yaml.
-    try:
-        service_account_info = json.loads(service_account_info_string)
-    except json.decoder.JSONDecodeError:
-        raise RuntimeError("gcp_credentials found in cluster yaml file but "
-                           "formatted improperly.")
-    gcp_credentials = service_account.Credentials.from_service_account_info(
-        service_account_info)
-    return gcp_credentials
+        return _create_crm_service_account(), \
+            _create_iam_service_account(), \
+            _create_compute_service_account()
+
+    assert ("type" in gcp_credentials and "credentials" in gcp_credentials), \
+        "gcp_credentials cluster yaml field missing 'type'" \
+        " or 'credentials' fields."
```

Comment:
Lets just break this up into two asserts so the user knows which one is missing.